// react-chat-engine/dist/index.modern.js
import React, { Component } from 'react';
import Websocket from 'react-websocket';
import _ from 'lodash';
import axios from 'axios';
import { newChat as newChat$1, isTyping as isTyping$1, sendMessage as sendMessage$1, removePerson as removePerson$1, addPerson as addPerson$1, getOtherPeople as getOtherPeople$1, deleteChat as deleteChat$1, editChat as editChat$1 } from 'react-chat-engine';
import { PlusOutlined, ArrowUpOutlined, DeleteOutlined, UserAddOutlined, CloseOutlined, LoadingOutlined, CloseCircleTwoTone, PaperClipOutlined, LeftOutlined, DownOutlined } from '@ant-design/icons';
import { setConfiguration, Row, Col } from 'react-grid-system';
import { animateScroll } from 'react-scroll';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

var Socket = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Socket, _Component);

  function Socket() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Socket.prototype;

  _proto.handleEvent = function handleEvent(event) {
    var props = this.props;
    var eventJSON = JSON.parse(event);

    if (eventJSON.action === 'is_typing') {
      props.onTyping && props.onTyping(eventJSON.data.id, eventJSON.data.person);
    } else if (eventJSON.action === 'new_chat') {
      props.onNewChat && props.onNewChat(eventJSON.data);
    } else if (eventJSON.action === 'edit_chat') {
      props.onEditChat && props.onEditChat(eventJSON.data);
    } else if (eventJSON.action === 'delete_chat') {
      props.onDeleteChat && props.onDeleteChat(eventJSON.data);
    } else if (eventJSON.action === 'add_person') {
      props.onAddPerson && props.onAddPerson(eventJSON.data);
    } else if (eventJSON.action === 'remove_person') {
      props.onRemovePerson && props.onRemovePerson(eventJSON.data);
    } else if (eventJSON.action === 'new_message') {
      props.onNewMessage && props.onNewMessage(eventJSON.data.id, eventJSON.data.message);
    } else if (eventJSON.action === 'edit_message') {
      props.onEditMessage && props.onEditMessage(eventJSON.data.id, eventJSON.data.message);
    } else if (eventJSON.action === 'delete_message') {
      props.onDeleteMessage && props.onDeleteMessage(eventJSON.data.id, eventJSON.data.message);
    }
  };

  _proto.onClose = function onClose() {
    this.props.onFailAuth && this.props.onFailAuth(this.props);
  };

  _proto.render = function render() {
    var _this = this;

    var _this$props = this.props,
        publicKey = _this$props.publicKey,
        projectID = _this$props.projectID,
        userName = _this$props.userName,
        userPassword = _this$props.userPassword,
        userSecret = _this$props.userSecret,
        development = _this$props.development;
    var wsStart = development ? 'ws://' : 'wss://';
    var rootHost = development ? '127.0.0.1:8000' : 'api.chatengine.io';
    var project = publicKey ? publicKey : projectID;
    var secret = userPassword ? userPassword : userSecret;
    return /*#__PURE__*/React.createElement(Websocket, {
      url: "" + wsStart + rootHost + "/person/?publicKey=" + project + "&username=" + userName + "&secret=" + secret,
      onOpen: function onOpen() {
        return _this.props.onConnect && _this.props.onConnect(_this.props);
      },
      onClose: this.onClose.bind(this),
      onMessage: this.handleEvent.bind(this)
    });
  };

  return Socket;
}(Component);

function getApiUrl(props) {
  if (props.development) {
    return 'http://127.0.0.1:8000';
  }

  return 'https://api.chatengine.io';
}

function getChats(props, callback) {
  axios.get(getApiUrl(props) + "/chats/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    props.onGetChats && props.onGetChats(response.data);
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Fetch Chats Error', error);
  });
}

function newChat(props, data, callback) {
  axios.post(getApiUrl(props) + "/chats/", data, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('New Chat Error', error);
  });
}

function editChat(props, chatId, data, callback) {
  axios.patch(getApiUrl(props) + "/chats/" + chatId + "/", data, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Edit Chat Error', error);
  });
}

function deleteChat(props, chatId, callback) {
  axios["delete"](getApiUrl(props) + "/chats/" + chatId + "/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Delete Chat Error', error);
  });
}

function getMessages(props, chatId, callback) {
  axios.get(getApiUrl(props) + "/chats/" + chatId + "/messages/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    props.onGetMessages && props.onGetMessages(chatId, response.data);
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Fetch Messages Error', error);
  });
}

function sendMessage(props, chatId, data, callback) {
  var formdata = new FormData();

  for (var i = 0; i < data.files.length; i++) {
    formdata.append('attachments', data.files[i], data.files[i].name);
  }

  formdata.append('text', data.text);
  axios.post(getApiUrl(props) + "/chats/" + chatId + "/messages/", formdata, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Send Messages Error', error);
  });
}

function editMessage(props, chatId, messageId, data, callback) {
  axios.patch(getApiUrl(props) + "/chats/" + chatId + "/messages/" + messageId + "/", data, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Delete Messages Error', error);
  });
}

function readMessage(props, chatId, messageId, callback) {
  axios.patch(getApiUrl(props) + "/chats/" + chatId + "/people/", {
    last_read: messageId
  }, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Send Messages Error', error);
  });
}

function deleteMessage(props, chatId, messageId, callback) {
  axios["delete"](getApiUrl(props) + "/chats/" + chatId + "/messages/" + messageId + "/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Delete Messages Error', error);
  });
}

function timeSinceDate(date) {
  var sent = Date(date);
  var day = sent.substr(0, 10);
  var time = sent.substr(15, 6);
  return day + " at " + time;
}
function daySinceSent(date) {
  var sent = Date(date);
  return sent.substr(4, 6);
}

var _dangerButton;

var Button = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Button, _Component);

  function Button() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      hovered: false
    };
    return _this;
  }

  var _proto = Button.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        value = _this$props.value,
        icon = _this$props.icon,
        theme = _this$props.theme,
        style = _this$props.style;
    var customStyle = style ? style : {};
    var hoverStyle = this.state.hovered ? styles.hoverButton : {};
    var themeStyle = theme == 'danger' ? styles.dangerButton : styles.button;
    return /*#__PURE__*/React.createElement("button", {
      type: this.props.type,
      onClick: function onClick() {
        return _this2.props.onClick && _this2.props.onClick();
      },
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          hovered: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          hovered: false
        });
      },
      style: _extends({}, themeStyle, customStyle, hoverStyle)
    }, icon == 'plus' && /*#__PURE__*/React.createElement(PlusOutlined, null), icon == 'send' && /*#__PURE__*/React.createElement(ArrowUpOutlined, null), icon == 'delete' && /*#__PURE__*/React.createElement(DeleteOutlined, null), icon == 'user-add' && /*#__PURE__*/React.createElement(UserAddOutlined, null), value && icon ? " " + value : value);
  };

  return Button;
}(Component);
var styles = {
  button: {
    color: 'white',
    border: 'none',
    outline: 'none',
    height: '36px',
    fontSize: '15px',
    cursor: 'pointer',
    padding: '8px 16px',
    borderRadius: '33px',
    backgroundColor: '#1890ff'
  },
  dangerButton: (_dangerButton = {
    color: 'red',
    border: 'none',
    outline: 'none',
    height: '36px',
    fontSize: '15px',
    cursor: 'pointer',
    padding: '8px 16px',
    borderRadius: '33px',
    backgroundColor: 'white'
  }, _dangerButton["border"] = '1px solid red', _dangerButton),
  hoverButton: {
    opacity: '0.66'
  }
};

var TextInput = /*#__PURE__*/function (_Component) {
  _inheritsLoose(TextInput, _Component);

  function TextInput() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      focused: false,
      value: null
    };
    return _this;
  }

  var _proto = TextInput.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var value = this.props["default"];

    if (value) {
      var event = {
        target: {
          value: value
        }
      };
      this.props.handleChange(event);
      this.setState({
        value: value
      });
    }
  };

  _proto.render = function render() {
    var _this2 = this;

    var customStyle = this.props.style ? this.props.style : {};
    var defaultStyle = this.state.focused ? styles$1.focusInput : styles$1.input;
    return (
      /*#__PURE__*/
      React.createElement("input", {
        value: this.props.value,
        placeholder: this.props.label,
        style: _extends({}, defaultStyle, customStyle),
        onBlur: function onBlur() {
          return _this2.setState({
            focused: false
          });
        },
        onFocus: function onFocus() {
          return _this2.setState({
            focused: true
          });
        },
        type: this.props.type ? this.props.type : "text",
        onChange: function onChange(e) {
          return _this2.props.handleChange && _this2.props.handleChange(e);
        }
      })
    );
  };

  return TextInput;
}(Component);
var styles$1 = {
  input: {
    height: '36px',
    fontSize: '15px',
    outline: 'none',
    borderRadius: '24px',
    border: '1px solid #d9d9d9',
    padding: '0px 12px',
    boxSizing: 'border-box'
  },
  focusInput: {
    height: '36px',
    fontSize: '15px',
    outline: 'none',
    borderRadius: '24px',
    border: '1px solid #1890ff',
    padding: '0px 12px',
    boxSizing: 'border-box'
  }
};

var TextAreaInput = /*#__PURE__*/function (_Component) {
  _inheritsLoose(TextAreaInput, _Component);

  function TextAreaInput() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      focused: false,
      value: null
    };
    return _this;
  }

  var _proto = TextAreaInput.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var value = this.props["default"];

    if (value) {
      this.props.handleChange({
        target: {
          value: value
        }
      });
      this.setState({
        value: value
      });
    }
  };

  _proto.resize = function resize() {
    var textarea = document.getElementById("msg-textarea");
    textarea.style.height = "";
    textarea.style.height = Math.min(textarea.scrollHeight, 150) + "px";
  };

  _proto.componentDidMount = function componentDidMount() {
    this.resize();
  };

  _proto.handleChange = function handleChange(e) {
    this.resize();
    this.props.handleChange && this.props.handleChange(e);
  };

  _proto.onKeyDown = function onKeyDown(e) {
    if (e.key === 'Enter') {
      e.preventDefault();

      if (this.props.value.length > 0) {
        this.props.onSubmit && this.props.onSubmit(e);
      }
    }
  };

  _proto.render = function render() {
    var _this2 = this;

    return /*#__PURE__*/React.createElement("textarea", {
      id: "msg-textarea",
      rows: "1",
      style: styles$2.input,
      value: this.props.value,
      placeholder: this.props.label,
      onBlur: function onBlur() {
        return _this2.setState({
          focused: false
        });
      },
      onFocus: function onFocus() {
        return _this2.setState({
          focused: true
        });
      },
      type: this.props.type ? this.props.type : "text",
      onChange: function onChange(e) {
        return _this2.handleChange(e);
      },
      onKeyDown: function onKeyDown(e) {
        return _this2.onKeyDown(e);
      }
    });
  };

  return TextAreaInput;
}(Component);
var styles$2 = {
  input: {
    border: '1px solid white',
    width: 'calc(100% - 64px - 24px - 44px)',
    outline: 'none',
    fontFamily: 'inherit',
    fontSize: '15px',
    paddingLeft: '12px',
    paddingRight: '12px',
    position: 'relative',
    left: '12px',
    resize: 'none'
  }
};

var AutoComplete = /*#__PURE__*/function (_Component) {
  _inheritsLoose(AutoComplete, _Component);

  function AutoComplete() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      focused: false,
      showAll: false
    };
    return _this;
  }

  var _proto = AutoComplete.prototype;

  _proto.onFocus = function onFocus() {
    this.onChange('', true);
    this.setState({
      focused: true
    });
    this.props.onFocus && this.props.onFocus();
  };

  _proto.onBlur = function onBlur() {
    this.setState({
      focused: false
    });
    this.props.onBlur && this.props.onBlur();
  };

  _proto.onChange = function onChange(value, showAll) {
    this.setState({
      showAll: showAll
    });
    this.props.handleChange && this.props.handleChange(value);
  };

  _proto.getNames = function getNames(options) {
    var _this2 = this;

    var max = 3;
    var count = 0;
    var results = [];
    options.map(function (option) {
      if (option.toLowerCase().indexOf(_this2.props.value.toLowerCase()) !== -1 && count < max) {
        count = count + 1;
        results.push(option);
      }
    });
    return results;
  };

  _proto.renderOptions = function renderOptions(options) {
    var _this3 = this;

    if (!this.props.value && !this.state.showAll) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    var results = this.getNames(options);
    return results.map(function (option, index) {
      return /*#__PURE__*/React.createElement("div", {
        key: "option_" + index
      }, _this3.props.renderOption && _this3.props.renderOption(option), index == results.length - 1 && /*#__PURE__*/React.createElement("div", {
        style: styles$3.close,
        onClick: function onClick() {
          return _this3.onChange('', false);
        }
      }, /*#__PURE__*/React.createElement(CloseOutlined, null)));
    });
  };

  _proto.render = function render() {
    var _this4 = this;

    var options = this.props.options;
    var customStyle = this.props.style ? this.props.style : {};

    var defaultStyle = _extends({}, styles$3.input, {
      border: this.state.focused ? '1px solid #1890ff' : '1px solid #d9d9d9'
    });

    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("input", {
      value: this.props.value,
      placeholder: this.props.label,
      style: _extends({}, defaultStyle, customStyle),
      type: this.props.type ? this.props.type : "text",
      onBlur: function onBlur() {
        return _this4.onBlur();
      },
      onFocus: function onFocus() {
        return _this4.onFocus();
      },
      onChange: function onChange(e) {
        return _this4.onChange(e.target.value, true);
      }
    }), options && options.length > 0 && this.state.showAll && /*#__PURE__*/React.createElement("div", {
      style: {
        borderRadius: '19px',
        border: '1px solid #afafaf'
      }
    }, this.renderOptions(this.props.options)));
  };

  return AutoComplete;
}(Component);
var styles$3 = {
  input: {
    height: '36px',
    fontSize: '15px',
    outline: 'none',
    borderRadius: '24px',
    padding: '0px 12px',
    boxSizing: 'border-box'
  },
  close: {
    cursor: 'pointer',
    textAlign: 'center',
    padding: '8px 12px',
    fontSize: '15px',
    borderRadius: '24px'
  }
};

var ChatForm = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(ChatForm, _React$Component);

  function ChatForm() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      value: ''
    };
    return _this;
  }

  var _proto = ChatForm.prototype;

  _proto.handleChange = function handleChange(event) {
    this.setState({
      value: event.target.value
    });
  };

  _proto.handleSubmit = function handleSubmit(event) {
    var _this2 = this;

    event.preventDefault();
    newChat$1(this.props.creds, {
      title: this.state.value
    }, function () {
      return _this2.setState({
        value: ''
      });
    });
  };

  _proto.render = function render() {
    var _this3 = this;

    return /*#__PURE__*/React.createElement("form", {
      onSubmit: this.handleSubmit.bind(this)
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: 'calc(100% - 26px)'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '1px'
      }
    }, /*#__PURE__*/React.createElement(TextInput, {
      label: "New Chat",
      value: this.state.value,
      style: {
        width: 'calc(100% - 48px)'
      },
      handleChange: function handleChange(e) {
        return _this3.handleChange(e);
      }
    })), /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        textAlign: 'right'
      }
    }, /*#__PURE__*/React.createElement(Button, {
      icon: "plus",
      type: "submit"
    }))));
  };

  return ChatForm;
}(React.Component);

var ChatList = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ChatList, _Component);

  function ChatList() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = ChatList.prototype;

  _proto.readLastMessage = function readLastMessage(creds, chat) {
    var readLastMessage = true;
    chat.people.map(function (chat_person) {
      if (creds && creds.userName === chat_person.person.username) {
        readLastMessage = chat.last_message.id === chat_person.last_read;
      }
    });
    return readLastMessage;
  };

  _proto.renderChats = function renderChats(chats) {
    var _this = this;

    return _.map(chats, function (chat, index) {
      var extraStyle = chat && _this.props.activeChat === chat.id ? styles$4.activeChat : {};
      if (!chat) return /*#__PURE__*/React.createElement("div", null);

      if (_this.props.renderChatCard) {
        return /*#__PURE__*/React.createElement("div", {
          key: "chat_" + index
        }, _this.props.renderChatCard(chat, index));
      }

      return /*#__PURE__*/React.createElement("div", {
        key: "chat_" + index,
        onClick: function onClick() {
          return _this.props.onChatClick(chat.id);
        },
        style: _extends({}, styles$4.chatContainer, extraStyle)
      }, /*#__PURE__*/React.createElement("div", {
        style: styles$4.titleText
      }, chat.title, ' ', !_this.readLastMessage(_this.props.creds, chat) && /*#__PURE__*/React.createElement("div", {
        style: {
          "float": 'right',
          width: '12px',
          height: '12px',
          borderRadius: '6px',
          backgroundColor: '#1890ff',
          display: 'inline-block'
        }
      })), /*#__PURE__*/React.createElement("div", {
        style: {
          width: '100%'
        }
      }, /*#__PURE__*/React.createElement("div", {
        style: styles$4.messageText
      }, chat.last_message.text ? chat.last_message.text : 'Say hello!'), /*#__PURE__*/React.createElement("div", {
        style: _extends({}, styles$4.messageText, {
          textAlign: 'right',
          width: '25%'
        })
      }, daySinceSent(chat.last_message.created))));
    });
  };

  _proto.render = function render() {
    var creds = this.props.creds;
    return /*#__PURE__*/React.createElement("div", {
      style: styles$4.chatListContainer
    }, /*#__PURE__*/React.createElement("div", {
      style: styles$4.chatsContainer
    }, this.renderChats(this.props.chats), /*#__PURE__*/React.createElement("div", {
      style: {
        height: '64px'
      }
    }), this.props.renderNewChatForm ? this.props.renderNewChatForm(creds) : /*#__PURE__*/React.createElement("div", {
      style: styles$4.newChatContainer
    }, /*#__PURE__*/React.createElement(ChatForm, {
      creds: creds
    }))));
  };

  return ChatList;
}(Component);

var styles$4 = {
  chatListContainer: {
    height: '100%',
    maxHeight: '100vh',
    overflow: 'scroll',
    borderRight: '1px solid #afafaf',
    backgroundColor: 'white'
  },
  chatsContainer: {
    width: '100%',
    backgroundColor: 'white',
    borderRadius: '0px 0px 24px 24px'
  },
  chatContainer: {
    padding: '16px',
    paddingBottom: '12px',
    cursor: 'pointer'
  },
  titleText: {
    fontWeight: '500',
    paddingBottom: '4px',
    whiteSpace: 'nowrap',
    overflow: 'hidden'
  },
  messageText: {
    width: '75%',
    color: 'rgba(153, 153, 153, 1)',
    fontSize: '14px',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    display: 'inline-block'
  },
  activeChat: {
    backgroundColor: '#d9d9d9',
    border: '4px solid white',
    borderRadius: '12px'
  },
  newChatContainer: {
    position: 'absolute',
    bottom: '0px',
    width: '100%',
    padding: '12px'
  }
};

var Title = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Title, _Component);

  function Title() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Title.prototype;

  _proto.render = function render() {
    var chat = this.props.chat;

    if (!chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: {
        position: 'absolute',
        top: '0px',
        width: '100%',
        zIndex: '999'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: styles$5.titleContainer
    }, /*#__PURE__*/React.createElement("div", {
      style: styles$5.titleText
    }, chat && chat.title), /*#__PURE__*/React.createElement("div", {
      style: styles$5.subtitleText
    }, "Active ", timeSinceDate(chat.last_message.created))));
  };

  return Title;
}(Component);
var styles$5 = {
  titleContainer: {
    padding: '18px',
    textAlign: 'center',
    color: 'rgb(24, 144, 255)',
    backgroundColor: 'rgb(256, 256, 256, 0.92)'
  },
  titleText: {
    fontSize: '24px',
    fontWeight: '600'
  },
  subtitleText: {
    fontSize: '12px'
  }
};

var AuthFail = /*#__PURE__*/function (_Component) {
  _inheritsLoose(AuthFail, _Component);

  function AuthFail() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = AuthFail.prototype;

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        height: '100%',
        textAlign: 'center'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%'
      }
    }, /*#__PURE__*/React.createElement("img", {
      style: {
        width: '40%',
        paddingTop: '25vh'
      },
      src: "https://chat-engine-assets.s3.amazonaws.com/denied.svg",
      alt: "welcome-to-chat-engine"
    })), /*#__PURE__*/React.createElement("div", {
      style: {
        fontSize: '18px',
        paddingTop: '24px',
        width: '75%',
        paddingLeft: '12.5%'
      }
    }, "Your credentials are incorrect. Make sure your Public Key, Username, and Password are correct at ", /*#__PURE__*/React.createElement("a", {
      href: "https://chatengine.io"
    }, "chatengine.io")));
  };

  return AuthFail;
}(Component);

var Loading = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Loading, _Component);

  function Loading() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Loading.prototype;

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement(LoadingOutlined, {
      style: {
        width: '100%',
        position: 'relative',
        top: 'calc(50% - 26px)',
        textAlign: 'center',
        fontSize: '52px'
      }
    });
  };

  return Loading;
}(Component);

var Welcome = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Welcome, _Component);

  function Welcome() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Welcome.prototype;

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        height: '100%',
        textAlign: 'center'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%'
      }
    }, /*#__PURE__*/React.createElement("img", {
      style: {
        width: '60%',
        paddingTop: '33vh'
      },
      src: "https://chat-engine-assets.s3.amazonaws.com/welcome.svg",
      alt: "welcome-to-chat-engine"
    })), /*#__PURE__*/React.createElement("div", {
      style: {
        fontSize: '18px',
        paddingTop: '24px',
        width: '75%',
        paddingLeft: '12.5%'
      }
    }, "Fill out the \"New Chat\" form to get started."));
  };

  return Welcome;
}(Component);

var months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
var suffix = ['st', 'nd', 'rd', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'st', 'nd', 'rd', 'th', 'th', 'th', 'th', 'th', 'th', 'th', 'st'];

var DatePartition = /*#__PURE__*/function (_Component) {
  _inheritsLoose(DatePartition, _Component);

  function DatePartition() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = DatePartition.prototype;

  _proto.getDate = function getDate(date) {
    return date ? date.substr(0, 10) : null;
  };

  _proto.formatDate = function formatDate(date_time) {
    var year = date_time.substr(0, 4);
    var month = months[parseInt(date_time.substr(5, 2)) - 1];
    var day = date_time.substr(8, 2);
    var dayDuffix = suffix[parseInt(date_time.substr(8, 2)) - 1];
    var time = date_time.substr(11, 5);
    var timeSuffix = '';

    if (parseInt(time.substr(0, 2)) >= 12) {
      if (parseInt(time.substr(0, 2)) > 12) {
        time = String(parseInt(time.substr(0, 2)) - 12) + ':' + time.substr(3, 2);
      }

      timeSuffix = 'PM';
    } else {
      timeSuffix = 'AM';
    }

    return time + timeSuffix + ', ' + month + ' ' + day + dayDuffix + ', ' + year;
  };

  _proto.render = function render() {
    var _this$props = this.props,
        lastCreated = _this$props.lastCreated,
        created = _this$props.created;
    var lastDate = this.getDate(lastCreated);
    var thisDate = this.getDate(created);

    if (lastCreated && lastDate === thisDate) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: styles$6.dateText
    }, this.formatDate(created));
  };

  return DatePartition;
}(Component);
var styles$6 = {
  dateText: {
    width: '100%',
    textAlign: 'center',
    paddingTop: '4px',
    paddingBottom: '10px',
    fontSize: '15px',
    color: 'rgba(0, 0, 0, .40)'
  }
};

var _thumbnail;

var Thumbnail = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Thumbnail, _Component);

  function Thumbnail() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Thumbnail.prototype;

  _proto.render = function render() {
    var attachment = this.props.attachment;
    return /*#__PURE__*/React.createElement("img", {
      style: styles$7.thumbnail,
      src: attachment.file,
      alt: 'image'
    });
  };

  return Thumbnail;
}(Component);
var styles$7 = {
  thumbnail: (_thumbnail = {
    width: '100%',
    cursor: 'pointer',
    textAlign: 'right',
    display: 'inline',
    objectFit: 'cover',
    borderRadius: '0.3em',
    paddingRight: '2px',
    height: '30vw'
  }, _thumbnail["width"] = '30vw', _thumbnail.maxHeight = '200px', _thumbnail.maxWidth = '200px', _thumbnail.minHeight = '100px', _thumbnail.minWidth = '100px', _thumbnail)
};

var colors = ['#D64045', '#5B3000', '#00CC99', '#467599', '#1D3354', '#8F250C', '#6153CC', '#961D4E', '#A29F15', '#0CAADC', '#FF5154', '#FA7921', '#688E26', '#550527', '#A10702', '#FF1053', '#6C6EA0', '#100B00'];

function stringToNumber(str) {
  var sum = 0;

  for (var i = 0; i < str.length; i++) {
    sum = sum + str.charCodeAt(i) * i - 97;
  }

  return sum;
}

function stringToColor(str) {
  if (!str) {
    return 'black';
  } else {
    return colors[stringToNumber(str) % colors.length];
  }
}

var Dot = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Dot, _Component);

  function Dot() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Dot.prototype;

  _proto.render = function render() {
    var customStyle = this.props.style ? this.props.style : {};
    var color = stringToColor(this.props.text);
    return /*#__PURE__*/React.createElement("div", {
      style: _extends({}, styles$8.avatar, customStyle, {
        backgroundColor: color
      })
    });
  };

  return Dot;
}(Component);
var styles$8 = {
  avatar: {
    width: '10px',
    height: '10px',
    borderRadius: '22px',
    textAlign: 'center'
  }
};

setConfiguration({
  maxScreenClass: 'xl'
});

var Message = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Message, _Component);

  function Message() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      selected: false
    };
    return _this;
  }

  var _proto = Message.prototype;

  _proto.renderReads = function renderReads() {
    var _this$props = this.props,
        chat = _this$props.chat,
        message = _this$props.message;

    if (!chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return chat.people.map(function (chatPerson, index) {
      if (message.id == chatPerson.last_read) {
        return /*#__PURE__*/React.createElement(Dot, {
          key: "read_" + index,
          text: chatPerson.person.username,
          style: {
            "float": 'right',
            marginLeft: '4px'
          }
        });
      }
    });
  };

  _proto.renderAttachments = function renderAttachments(borderRadius) {
    var attachments = this.props.message ? this.props.message.attachments : [];
    return attachments.map(function (attachment, index) {
      return /*#__PURE__*/React.createElement(Thumbnail, {
        attachment: attachment,
        key: "attachment_" + index,
        borderRadius: borderRadius
      });
    });
  };

  _proto.render = function render() {
    var _this2 = this;

    var _this$props2 = this.props,
        lastMessage = _this$props2.lastMessage,
        message = _this$props2.message,
        nextMessage = _this$props2.nextMessage;

    if (!message) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    var topRightRadius = !lastMessage || lastMessage.sender.username !== message.sender.username ? '1.3em' : '0.3em';
    var bottomRightRadius = !nextMessage || nextMessage.sender.username !== message.sender.username ? '1.3em' : '0.3em';
    var borderRadius = "1.3em " + topRightRadius + " " + bottomRightRadius + " 1.3em";
    var paddingBottom = !nextMessage || nextMessage.sender.username !== message.sender.username ? '12px' : '2px';
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        textAlign: 'right',
        paddingBottom: paddingBottom
      },
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          selected: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          selected: false
        });
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        display: 'auto'
      }
    }, this.renderAttachments()), /*#__PURE__*/React.createElement(Row, {
      style: {
        paddingRight: '2px'
      }
    }, /*#__PURE__*/React.createElement(Col, {
      xs: 1,
      sm: 2,
      md: 3
    }), /*#__PURE__*/React.createElement(Col, {
      xs: 11,
      sm: 10,
      md: 9
    }, /*#__PURE__*/React.createElement("div", {
      style: _extends({}, styles$9.myMessage, {
        borderRadius: borderRadius
      })
    }, message.text)), /*#__PURE__*/React.createElement(Col, {
      xs: 1,
      sm: 2,
      md: 3
    }), /*#__PURE__*/React.createElement(Col, {
      xs: 12
    }, this.renderReads())));
  };

  return Message;
}(Component);
var styles$9 = {
  myMessage: {
    color: 'white',
    cursor: 'pointer',
    "float": 'right',
    textAlign: 'left',
    padding: '12px',
    fontSize: '15px',
    whiteSpace: 'pre-line',
    backgroundColor: 'rgb(24, 144, 255)'
  }
};

var Avatar = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Avatar, _Component);

  function Avatar() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Avatar.prototype;

  _proto.render = function render() {
    var person = this.props.person;
    var customStyle = this.props.style ? this.props.style : {};
    var text = person.username ? person.username.substring(0, 2).toUpperCase() : '';
    var color = stringToColor(person ? person.username : '');
    return /*#__PURE__*/React.createElement("div", null, this.props.show_online !== false && /*#__PURE__*/React.createElement("div", {
      style: {
        width: '8px',
        height: '8px',
        borderRadius: '100%',
        position: 'absolute',
        border: '2px solid white',
        backgroundColor: person.is_online ? '#52c41a' : '#f5222d'
      }
    }), /*#__PURE__*/React.createElement("div", {
      style: _extends({}, styles$a.avatar, customStyle, {
        backgroundColor: color
      })
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        color: 'white',
        paddingTop: '12px',
        fontSize: '15px',
        fontWeight: '600'
      }
    }, text)));
  };

  return Avatar;
}(Component);
var styles$a = {
  avatar: {
    width: '44px',
    height: '44px',
    borderRadius: '22px',
    color: 'white',
    textAlign: 'center'
  }
};

setConfiguration({
  maxScreenClass: 'xl'
});

var TheirMessage = /*#__PURE__*/function (_Component) {
  _inheritsLoose(TheirMessage, _Component);

  function TheirMessage() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      selected: false
    };
    return _this;
  }

  var _proto = TheirMessage.prototype;

  _proto.renderReads = function renderReads() {
    var _this$props = this.props,
        chat = _this$props.chat,
        message = _this$props.message;

    if (!chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return chat.people.map(function (person, index) {
      if (message.id == person.last_read) {
        return /*#__PURE__*/React.createElement(Dot, {
          key: "read_" + index,
          text: person.person.username,
          style: {
            "float": 'left',
            marginLeft: '4px'
          }
        });
      }
    });
  };

  _proto.renderAttachments = function renderAttachments(borderRadius) {
    var attachments = this.props.message ? this.props.message.attachments : [];
    return attachments.map(function (attachment, index) {
      return /*#__PURE__*/React.createElement(Thumbnail, {
        attachment: attachment,
        key: "attachment_" + index,
        borderRadius: borderRadius
      });
    });
  };

  _proto.render = function render() {
    var _this2 = this;

    var _this$props2 = this.props,
        lastMessage = _this$props2.lastMessage,
        message = _this$props2.message,
        nextMessage = _this$props2.nextMessage;

    if (!message) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    var topLeftRadius = !lastMessage || lastMessage.sender.username !== message.sender.username ? '1.3em' : '0.3em';
    var bottomLeftRadius = !nextMessage || nextMessage.sender.username !== message.sender.username ? '1.3em' : '0.3em';
    var borderRadius = topLeftRadius + " 1.3em 1.3em " + bottomLeftRadius;
    var paddingBottom = !nextMessage || nextMessage.sender.username !== message.sender.username ? '12px' : '2px';
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        paddingBottom: paddingBottom
      }
    }, (!lastMessage || lastMessage.sender.username !== message.sender.username) && /*#__PURE__*/React.createElement("div", {
      style: styles$b.nameText
    }, message.sender.username), /*#__PURE__*/React.createElement("div", {
      style: {
        display: 'auto',
        paddingLeft: '50px'
      }
    }, this.renderAttachments()), /*#__PURE__*/React.createElement(Row, {
      style: {
        paddingLeft: '2px'
      }
    }, /*#__PURE__*/React.createElement(Col, {
      xs: 11,
      sm: 10,
      md: 9
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '0px'
      }
    }, !nextMessage || nextMessage.sender.username !== message.sender.username ? /*#__PURE__*/React.createElement(Avatar, {
      person: message.sender
    }) : /*#__PURE__*/React.createElement("div", null)), /*#__PURE__*/React.createElement("div", {
      style: {
        paddingLeft: '48px'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: _extends({}, styles$b.theirMessage, {
        borderRadius: borderRadius
      }),
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          selected: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          selected: false
        });
      }
    }, message.text))), /*#__PURE__*/React.createElement(Col, {
      xs: 9,
      style: {
        marginLeft: '48px'
      }
    }, this.renderReads())));
  };

  return TheirMessage;
}(Component);
var styles$b = {
  theirMessage: {
    cusor: 'auto',
    color: 'black',
    "float": 'left',
    padding: '12px',
    fontSize: '15px',
    whiteSpace: 'pre-line',
    backgroundColor: '#f1f0f0'
  },
  nameText: {
    paddingLeft: '62px',
    paddingBottom: '2px',
    color: 'rgba(0, 0, 0, .40)',
    fontSize: '15px'
  }
};

var Message$1 = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Message$1, _Component);

  function Message$1() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      selected: false
    };
    return _this;
  }

  var _proto = Message$1.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        creds = _this$props.creds,
        chat = _this$props.chat,
        lastMessage = _this$props.lastMessage,
        message = _this$props.message,
        nextMessage = _this$props.nextMessage;

    if (!message) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(DatePartition, {
      lastCreated: lastMessage ? lastMessage.created : null,
      created: message.created
    }), message.sender.username == creds.userName ? /*#__PURE__*/React.createElement(Message, {
      chat: chat,
      creds: creds,
      lastMessage: lastMessage,
      message: message,
      nextMessage: nextMessage
    }) : /*#__PURE__*/React.createElement(TheirMessage, {
      chat: chat,
      creds: creds,
      lastMessage: lastMessage,
      message: message,
      nextMessage: nextMessage
    }));
  };

  return Message$1;
}(Component);

var FileRow = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(FileRow, _React$Component);

  function FileRow() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      hovered: false
    };
    return _this;
  }

  var _proto = FileRow.prototype;

  _proto.render = function render() {
    var _this2 = this;

    return /*#__PURE__*/React.createElement("div", {
      style: {
        padding: '12px 6px',
        display: 'inline'
      },
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          hovered: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          hovered: false
        });
      }
    }, /*#__PURE__*/React.createElement("img", {
      style: styles$c.imageSquare,
      alt: this.props.file ? this.props.file.name : '',
      src: URL.createObjectURL(this.props.file)
    }), this.state.hovered && /*#__PURE__*/React.createElement(CloseCircleTwoTone, {
      style: styles$c.closeIcon,
      onClick: function onClick() {
        return _this2.props.onRemove && _this2.props.onRemove();
      }
    }));
  };

  return FileRow;
}(React.Component);
var styles$c = {
  imageSquare: {
    height: '108px',
    width: '108px',
    border: '1px solid #afafaf',
    borderRadius: '8px',
    objectFit: 'cover',
    display: 'inline'
  },
  closeIcon: {
    position: 'relative',
    bottom: '91px',
    right: '20px',
    width: '0px',
    cursor: 'pointer'
  }
};

var FileRow$1 = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(FileRow$1, _React$Component);

  function FileRow$1() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = FileRow$1.prototype;

  _proto.renderFiles = function renderFiles() {
    var _this = this;

    return this.props.files.map(function (file, index) {
      return /*#__PURE__*/React.createElement(FileRow, {
        file: file,
        key: "thumb_" + index,
        onRemove: function onRemove() {
          return _this.props.onRemove && _this.props.onRemove(index);
        }
      });
    });
  };

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: 'calc(100% - 24px)',
        paddingLeft: '12px'
      }
    }, this.renderFiles());
  };

  return FileRow$1;
}(React.Component);

var ImagesInput = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(ImagesInput, _React$Component);

  function ImagesInput() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      hovered: false,
      files: []
    };
    return _this;
  }

  var _proto = ImagesInput.prototype;

  _proto.onSelect = function onSelect(event) {
    var files = [];
    var indexes = [].concat(Array(event.target.files.length).keys());
    indexes.map(function (i, index) {
      var file = event.target.files[index];

      if (file) {
        files.push(file);
      }
    });
    this.props.onSelectFiles && this.props.onSelectFiles(files);
  };

  _proto.render = function render() {
    var _this2 = this;

    return /*#__PURE__*/React.createElement("form", {
      className: "uploader",
      encType: "multipart/form-data",
      style: {
        height: '0px'
      }
    }, /*#__PURE__*/React.createElement("label", {
      htmlFor: "files",
      id: "upload-document-button"
    }, /*#__PURE__*/React.createElement(PaperClipOutlined, {
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          hovered: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          hovered: false
        });
      },
      style: _extends({}, {
        cursor: 'pointer',
        position: 'absolute',
        bottom: '20px',
        right: '63px',
        fontSize: '18px'
      }, {
        color: this.state.hovered ? '#69c0ff' : '#1890ff'
      })
    })), /*#__PURE__*/React.createElement("input", {
      multiple: true,
      id: "files",
      accept: "image/x-png,image/gif,image/jpeg",
      style: {
        visibility: "hidden"
      },
      type: "file",
      onChange: function onChange(e) {
        return _this2.onSelect(e);
      }
    }));
  };

  return ImagesInput;
}(React.Component);

var MessageForm = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(MessageForm, _React$Component);

  function MessageForm() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      value: '',
      files: []
    };
    return _this;
  }

  var _proto = MessageForm.prototype;

  _proto.onRemove = function onRemove(index) {
    var files = this.state.files;
    files.splice(index, 1);
    this.setState({
      files: files
    });
  };

  _proto.handleChange = function handleChange(event) {
    this.setState({
      value: event.target.value
    });
    isTyping$1(this.props.creds, this.props.chatId);
  };

  _proto.handleSubmit = function handleSubmit(event) {
    var _this2 = this;

    event.preventDefault();
    sendMessage$1(this.props.creds, this.props.chatId, {
      text: this.state.value,
      files: this.state.files
    }, function (data) {
      return _this2.setState({
        value: '',
        files: []
      });
    });
  };

  _proto.renderFiles = function renderFiles() {
    return this.state.files.map(function (file, index) {
      return /*#__PURE__*/React.createElement("div", {
        key: "file_" + index
      }, /*#__PURE__*/React.createElement("img", {
        style: {
          maxHeight: '66px',
          maxWidth: '66px'
        },
        src: URL.createObjectURL(file),
        alt: file.name
      }));
    });
  };

  _proto.render = function render() {
    var _this3 = this;

    return /*#__PURE__*/React.createElement("div", {
      style: styles$d.messageFormContainer,
      id: "msg-form-container"
    }, /*#__PURE__*/React.createElement(FileRow$1, {
      files: this.state.files,
      onRemove: function onRemove(i) {
        return _this3.onRemove(i);
      }
    }), /*#__PURE__*/React.createElement(ImagesInput, {
      onSelectFiles: function onSelectFiles(files) {
        return _this3.setState({
          files: files
        });
      }
    }), /*#__PURE__*/React.createElement("form", {
      onSubmit: this.handleSubmit.bind(this)
    }, /*#__PURE__*/React.createElement("div", {
      style: styles$d.inputContainer
    }, /*#__PURE__*/React.createElement(TextAreaInput, {
      label: "Send a message...",
      value: this.state.value,
      handleChange: this.handleChange.bind(this),
      onSubmit: this.handleSubmit.bind(this)
    }), /*#__PURE__*/React.createElement(Button, {
      type: "submit",
      icon: "send",
      style: {
        position: 'absolute',
        bottom: '10px',
        right: '6px'
      }
    }))));
  };

  return MessageForm;
}(React.Component);
var styles$d = {
  messageFormContainer: {
    position: 'absolute',
    bottom: '0px',
    width: '100%',
    backgroundColor: 'white'
  },
  inputContainer: {
    overflow: 'scroll',
    minHeight: '36px',
    paddingTop: '10px',
    paddingBottom: '6px'
  }
};

var ChatList$1 = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ChatList, _Component);

  function ChatList() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      duration: 0
    };
    return _this;
  }

  var _proto = ChatList.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this2 = this;

    setTimeout(function () {
      _this2.setState({
        duration: 100
      });
    }, 3000);
  };

  _proto.renderTypers = function renderTypers() {
    var _this$props = this.props,
        typingData = _this$props.typingData,
        chatId = _this$props.chatId;
    var typers = typingData[chatId] ? typingData[chatId] : [];

    if (this.props.renderIsTyping) {
      return this.props.renderIsTyping(typers);
    }

    return typers.map(function (typer, index) {
      return /*#__PURE__*/React.createElement("div", {
        key: "typer_" + index,
        style: {
          color: stringToColor(typer),
          padding: '2px',
          paddingLeft: '12px'
        }
      }, typer + " is typing...");
    });
  };

  _proto.renderMessages = function renderMessages() {
    var _this3 = this;

    var _this$props2 = this.props,
        messages = _this$props2.messages,
        chats = _this$props2.chats,
        creds = _this$props2.creds,
        chatId = _this$props2.chatId;
    var chat = chats && chats[chatId];
    var keys = Object.keys(messages);
    return keys.map(function (key, index) {
      var message = messages[key];
      var lastMessageKey = index == 0 ? null : keys[index - 1];
      var nextMessageKey = index == keys.length - 1 ? null : keys[index + 1];

      if (_this3.props.renderMessageBubble) {
        return /*#__PURE__*/React.createElement("div", {
          key: "message_" + index
        }, _this3.props.renderMessageBubble(creds, chat, messages[lastMessageKey], message, messages[nextMessageKey]));
      }

      return /*#__PURE__*/React.createElement(Message$1, {
        key: "message_" + index,
        chat: chat,
        creds: creds,
        message: message,
        lastMessage: messages[lastMessageKey],
        nextMessage: messages[nextMessageKey]
      });
    });
  };

  _proto.scrollToBottom = function scrollToBottom() {
    animateScroll.scrollToBottom({
      duration: this.state.duration,
      containerId: "feed-container"
    });
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    this.scrollToBottom();
  };

  _proto.render = function render() {
    var _this$props3 = this.props,
        chats = _this$props3.chats,
        creds = _this$props3.creds,
        chatId = _this$props3.chatId;
    var chat = chats && chats[chatId];

    if (creds === null) {
      return /*#__PURE__*/React.createElement(Loading, null);
    }

    if (creds === undefined) {
      return /*#__PURE__*/React.createElement(AuthFail, null);
    }

    if (creds && chats !== null && _.isEmpty(chats)) {
      return /*#__PURE__*/React.createElement(Welcome, null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: {
        display: 'flex',
        maxHeight: '100vh',
        backgroundColor: '#f0f0f0'
      }
    }, this.props.renderChatHeader ? this.props.renderChatHeader(chat) : /*#__PURE__*/React.createElement(Title, {
      chat: chat
    }), /*#__PURE__*/React.createElement("div", {
      style: styles$e.feedContainer,
      id: "feed-container"
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '88px'
      }
    }), this.renderMessages(), this.renderTypers(), /*#__PURE__*/React.createElement("div", {
      id: "feet-bottom",
      style: {
        height: '54px'
      }
    })), this.props.renderNewMessageForm ? this.props.renderNewMessageForm(creds, chatId) : /*#__PURE__*/React.createElement(MessageForm, {
      chatId: chatId,
      creds: creds
    }));
  };

  return ChatList;
}(Component);
var styles$e = {
  feedContainer: {
    position: 'absolute',
    top: '0px',
    height: '100%',
    width: '100%',
    overflow: 'scroll',
    backgroundColor: 'white'
  }
};

var Person = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Person, _Component);

  function Person() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      selected: false
    };
    return _this;
  }

  var _proto = Person.prototype;

  _proto.onRemovePerson = function onRemovePerson() {
    removePerson$1(this.props.creds, this.props.chat.id, this.props.person.username);
  };

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        person = _this$props.person,
        chat = _this$props.chat,
        creds = _this$props.creds;

    if (!person || !chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    var admin = chat.admin;
    return /*#__PURE__*/React.createElement("div", {
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          selected: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          selected: false
        });
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        padding: '2px',
        height: '0px'
      }
    }, /*#__PURE__*/React.createElement(Avatar, {
      person: person
    })), /*#__PURE__*/React.createElement("div", {
      style: {
        paddingLeft: '52px',
        height: '44px',
        position: 'relative',
        top: '10px',
        fontSize: '15px'
      }
    }, person.username !== admin.username ? person.username : person.username + " (Admin)"), this.state.selected && creds.userName === admin.username && person.username !== admin.username && /*#__PURE__*/React.createElement("div", {
      style: {
        "float": 'right',
        height: '0px',
        position: 'relative',
        bottom: '44px'
      }
    }, /*#__PURE__*/React.createElement(Button, {
      theme: "danger",
      icon: "delete",
      onClick: function onClick() {
        return _this2.onRemovePerson();
      }
    })));
  };

  return Person;
}(Component);

var Option = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Option, _React$Component);

  function Option() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      focused: false
    };
    return _this;
  }

  var _proto = Option.prototype;

  _proto.render = function render() {
    var _this2 = this;

    return /*#__PURE__*/React.createElement("div", {
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          focused: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          focused: false
        });
      },
      onClick: function onClick() {
        return _this2.props.onClick && _this2.props.onClick();
      },
      style: _extends({}, styles$f.option, {
        backgroundColor: this.state.focused ? '#f5f5f5' : 'white'
      })
    }, this.props.text);
  };

  return Option;
}(React.Component);
var styles$f = {
  option: {
    borderRadius: '24px',
    padding: '8px 16px',
    cursor: 'pointer',
    fontSize: '15px'
  }
};

var PersonForm = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(PersonForm, _React$Component);

  function PersonForm() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      value: '',
      others: []
    };
    return _this;
  }

  var _proto = PersonForm.prototype;

  _proto.handleChange = function handleChange(value) {
    this.setState({
      value: value
    });
  };

  _proto.addPerson = function addPerson(name) {
    var _this2 = this;

    addPerson$1(this.props.creds, this.props.chat.id, name, function () {
      _this2.setState({
        value: ''
      });

      _this2.getOthers();
    });
  };

  _proto.onGetOthers = function onGetOthers(others) {
    var usernames = [];
    others.map(function (other) {
      usernames.push(other.username);
    });
    this.setState({
      others: usernames
    });
  };

  _proto.getOthers = function getOthers() {
    var _this3 = this;

    getOtherPeople$1(this.props.creds, this.props.chat.id, function (id, others) {
      return _this3.onGetOthers(others);
    }, function () {});
  };

  _proto.renderOption = function renderOption(option) {
    var _this4 = this;

    return /*#__PURE__*/React.createElement(Option, {
      text: option,
      onClick: function onClick() {
        return _this4.addPerson(option);
      }
    });
  };

  _proto.render = function render() {
    var _this5 = this;

    return /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: '12px'
      }
    }, /*#__PURE__*/React.createElement(AutoComplete, {
      style: {
        width: '100%'
      },
      label: "Type a username",
      value: this.state.value,
      options: this.state.others,
      onFocus: function onFocus() {
        return _this5.getOthers();
      },
      handleChange: function handleChange(value) {
        return _this5.handleChange(value);
      },
      renderOption: function renderOption(option) {
        return _this5.renderOption(option);
      }
    }));
  };

  return PersonForm;
}(React.Component);

var PeopleList = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(PeopleList, _React$Component);

  function PeopleList() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      collapsed: false,
      hovered: false
    };
    return _this;
  }

  var _proto = PeopleList.prototype;

  _proto.renderChatPeople = function renderChatPeople(people) {
    var _this2 = this;

    return people.map(function (chatPerson, index) {
      return /*#__PURE__*/React.createElement(Person, {
        key: "person_" + index,
        person: chatPerson.person,
        creds: _this2.props.creds,
        chat: _this2.props.chat
      });
    });
  };

  _proto.render = function render() {
    var _this3 = this;

    var _this$props = this.props,
        creds = _this$props.creds,
        chat = _this$props.chat;

    if (!chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: {
        borderTop: '1px solid #f0f0f0'
      }
    }, /*#__PURE__*/React.createElement("div", {
      onMouseEnter: function onMouseEnter() {
        return _this3.setState({
          hovered: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this3.setState({
          hovered: false
        });
      },
      onClick: function onClick() {
        return _this3.setState({
          collapsed: !_this3.state.collapsed
        });
      },
      style: this.state.hovered ? {
        backgroundColor: '#f0f0f0',
        cursor: 'pointer'
      } : {}
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        fontSize: '17px',
        padding: '12px',
        fontWeight: '600'
      }
    }, "People"), this.state.collapsed ? /*#__PURE__*/React.createElement(LeftOutlined, {
      style: styles$g.collapseIcon
    }) : /*#__PURE__*/React.createElement(DownOutlined, {
      style: styles$g.collapseIcon
    })), !this.state.collapsed && /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '12px'
      }
    }), this.renderChatPeople(chat.people), /*#__PURE__*/React.createElement("div", {
      style: {
        height: '12px'
      }
    }), creds && chat && creds.userName === chat.admin.username && /*#__PURE__*/React.createElement(PersonForm, {
      creds: creds,
      chat: chat
    })));
  };

  return PeopleList;
}(React.Component);
var styles$g = {
  collapseIcon: {
    "float": 'right',
    position: 'relative',
    bottom: '30px',
    right: '12px'
  }
};

var Thumbnail$1 = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Thumbnail, _React$Component);

  function Thumbnail() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Thumbnail.prototype;

  _proto.render = function render() {
    var attachment = this.props.attachment;

    if (!attachment) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: styles$h.container
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        paddingTop: '100%'
      }
    }), /*#__PURE__*/React.createElement("img", {
      alt: attachment.id,
      style: styles$h.image,
      src: attachment.file,
      id: "image-" + attachment.id
    }));
  };

  return Thumbnail;
}(React.Component);
var styles$h = {
  container: {
    position: 'relative',
    width: 'calc(33% - 2px)',
    border: '1px solid white',
    display: 'inline-block',
    cursor: 'pointer'
  },
  image: {
    top: '0px',
    width: '100%',
    height: '100%',
    position: 'absolute',
    objectFit: 'cover'
  }
};

var Photos = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Photos, _React$Component);

  function Photos() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      collapsed: true,
      hovered: false
    };
    return _this;
  }

  var _proto = Photos.prototype;

  _proto.renderPhotos = function renderPhotos(attachments) {
    return attachments.map(function (attachment, index) {
      return /*#__PURE__*/React.createElement(Thumbnail$1, {
        key: "person_" + index,
        attachment: attachment
      });
    });
  };

  _proto.render = function render() {
    var _this2 = this;

    var chat = this.props.chat;

    if (!chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: {
        borderTop: '1px solid #f0f0f0'
      }
    }, /*#__PURE__*/React.createElement("div", {
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          hovered: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          hovered: false
        });
      },
      onClick: function onClick() {
        return _this2.setState({
          collapsed: !_this2.state.collapsed
        });
      },
      style: this.state.hovered ? {
        backgroundColor: '#f0f0f0',
        cursor: 'pointer'
      } : {}
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        fontSize: '17px',
        padding: '12px',
        fontWeight: '600'
      }
    }, "Photos"), this.state.collapsed ? /*#__PURE__*/React.createElement(LeftOutlined, {
      style: styles$i.collapseIcon
    }) : /*#__PURE__*/React.createElement(DownOutlined, {
      style: styles$i.collapseIcon
    })), !this.state.collapsed && /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '12px'
      }
    }), this.renderPhotos(chat.attachments)));
  };

  return Photos;
}(React.Component);
var styles$i = {
  collapseIcon: {
    "float": 'right',
    position: 'relative',
    bottom: '30px',
    right: '12px'
  }
};

var Options = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Options, _React$Component);

  function Options() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      collapsed: true,
      hovered: false
    };
    return _this;
  }

  var _proto = Options.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        chat = _this$props.chat,
        creds = _this$props.creds;

    if (!chat) {
      return /*#__PURE__*/React.createElement("div", null);
    }

    return /*#__PURE__*/React.createElement("div", {
      style: {
        borderTop: '1px solid #f0f0f0'
      }
    }, /*#__PURE__*/React.createElement("div", {
      onMouseEnter: function onMouseEnter() {
        return _this2.setState({
          hovered: true
        });
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.setState({
          hovered: false
        });
      },
      onClick: function onClick() {
        return _this2.setState({
          collapsed: !_this2.state.collapsed
        });
      },
      style: this.state.hovered ? {
        backgroundColor: '#f0f0f0',
        cursor: 'pointer'
      } : {}
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        fontSize: '17px',
        padding: '12px',
        fontWeight: '600'
      }
    }, "Options"), this.state.collapsed ? /*#__PURE__*/React.createElement(LeftOutlined, {
      style: styles$j.collapseIcon
    }) : /*#__PURE__*/React.createElement(DownOutlined, {
      style: styles$j.collapseIcon
    })), !this.state.collapsed && /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '12px'
      }
    }), /*#__PURE__*/React.createElement(Button, {
      value: "Delete",
      theme: "danger",
      icon: "delete",
      onClick: function onClick() {
        return deleteChat$1(creds, chat.id, function (data) {});
      },
      style: {
        width: '100%',
        marginBottom: '12px'
      }
    })));
  };

  return Options;
}(React.Component);
var styles$j = {
  collapseIcon: {
    "float": 'right',
    position: 'relative',
    bottom: '30px',
    right: '12px'
  }
};

var MessageForm$1 = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(MessageForm, _React$Component);

  function MessageForm() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.state = {
      activeChat: null,
      value: ''
    };
    return _this;
  }

  var _proto = MessageForm.prototype;

  _proto.handleChange = function handleChange(event) {
    this.setState({
      value: event.target.value
    });
  };

  _proto.handleSubmit = function handleSubmit(event) {
    event.preventDefault();
    editChat$1(this.props.creds, this.props.chat.id, {
      title: this.state.value
    }, function (data) {});
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    if (this.props.chat.title !== this.state.value && this.state.activeChat !== this.props.chat.id) {
      this.setState({
        value: this.props.chat.title,
        activeChat: this.props.chat.id
      });
    }
  };

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement("form", {
      onSubmit: this.handleSubmit.bind(this)
    }, /*#__PURE__*/React.createElement(TextInput, {
      label: "Rename this Chat",
      value: this.state.value,
      "default": this.props.chat.title,
      handleChange: this.handleChange.bind(this),
      style: {
        fontSize: '16px',
        fontWeight: '600',
        textAlign: 'center',
        border: '0px solid white',
        width: 'calc(100% - 24px)'
      }
    }));
  };

  return MessageForm;
}(React.Component);

var ChatSettings = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ChatSettings, _Component);

  function ChatSettings() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = ChatSettings.prototype;

  _proto.renderOnePerson = function renderOnePerson(people) {
    return /*#__PURE__*/React.createElement(Col, {
      xs: 6,
      style: {
        paddingTop: '22px',
        height: '80px'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        margin: 'auto',
        width: '33%'
      }
    }, /*#__PURE__*/React.createElement(Avatar, {
      show_online: false,
      person: people[0].person
    })));
  };

  _proto.renderTwoPeople = function renderTwoPeople(people) {
    return /*#__PURE__*/React.createElement(Col, {
      xs: 6,
      style: {
        paddingTop: '22px',
        height: '80px'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        margin: 'auto',
        width: '50%'
      }
    }, /*#__PURE__*/React.createElement(Avatar, {
      show_online: false,
      person: people[0].person,
      style: {
        "float": 'left',
        position: 'relative',
        right: '5px'
      }
    }), /*#__PURE__*/React.createElement(Avatar, {
      show_online: false,
      person: people[1].person,
      style: {
        "float": 'right',
        position: 'relative',
        left: '5px',
        bottom: '44px',
        zIndex: 11
      }
    })));
  };

  _proto.renderThreePeople = function renderThreePeople(people) {
    return /*#__PURE__*/React.createElement(Col, {
      xs: 6,
      style: {
        paddingTop: '22px',
        height: '80px'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        margin: 'auto',
        width: '50%'
      }
    }, /*#__PURE__*/React.createElement(Avatar, {
      show_online: false,
      person: people[0].person,
      style: {
        "float": 'right',
        position: 'relative',
        right: '28px',
        top: '10px',
        zIndex: 11
      }
    }), /*#__PURE__*/React.createElement(Avatar, {
      show_online: false,
      person: people[1].person,
      style: {
        "float": 'right',
        position: 'relative',
        right: '14px',
        bottom: '56px',
        zIndex: 11
      }
    }), /*#__PURE__*/React.createElement(Avatar, {
      show_online: false,
      person: people[2].person,
      style: {
        "float": 'right',
        position: 'relative',
        left: '5px',
        bottom: '78px',
        zIndex: 11
      }
    })));
  };

  _proto.render = function render() {
    var _this$props = this.props,
        chat = _this$props.chat,
        creds = _this$props.creds;
    var topPeople = chat ? chat.people.slice(0, 3) : [];
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Col, {
      xs: 3
    }), topPeople.length == 1 && this.renderOnePerson(topPeople), topPeople.length == 2 && this.renderTwoPeople(topPeople), topPeople.length == 3 && this.renderThreePeople(topPeople), /*#__PURE__*/React.createElement(Col, {
      xs: 3
    })), /*#__PURE__*/React.createElement(MessageForm$1, {
      chat: chat,
      creds: creds
    }));
  };

  return ChatSettings;
}(Component);

var ChatSettingsContainer = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ChatSettingsContainer, _Component);

  function ChatSettingsContainer() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = ChatSettingsContainer.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        chat = _this$props.chat,
        creds = _this$props.creds;
    if (!chat) return /*#__PURE__*/React.createElement("div", {
      style: styles$k.filler
    });
    return /*#__PURE__*/React.createElement("div", {
      style: styles$k.settingsContainer
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '90%',
        paddingLeft: '5%'
      }
    }, this.props.renderChatSettings ? this.props.renderChatSettings(creds, chat) : /*#__PURE__*/React.createElement(ChatSettings, {
      creds: creds,
      chat: chat
    }), this.props.renderPeopleSettings ? this.props.renderPeopleSettings(creds, chat) : /*#__PURE__*/React.createElement(PeopleList, {
      creds: creds,
      chat: chat
    }), this.props.renderPhotosSettings ? this.props.renderPhotosSettings(chat) : /*#__PURE__*/React.createElement(Photos, {
      creds: creds,
      chat: chat
    }), creds && chat && creds.userName === chat.admin.username && /*#__PURE__*/React.createElement("div", null, this.props.renderOptionsSettings ? this.props.renderOptionsSettings(creds, chat) : /*#__PURE__*/React.createElement(Options, {
      creds: creds,
      chat: chat
    }))));
  };

  return ChatSettingsContainer;
}(Component);
var styles$k = {
  settingsContainer: {
    height: '100%',
    overflow: 'scroll',
    borderLeft: '1px solid #afafaf',
    backgroundColor: 'white'
  },
  filler: {
    display: 'flex',
    width: '90%',
    paddingLeft: '5%',
    borderLeft: '1px solid #afafaf'
  }
};

setConfiguration({
  maxScreenClass: 'xl',
  gutterWidth: 0
});

var App = /*#__PURE__*/function (_Component) {
  _inheritsLoose(App, _Component);

  function App() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this.state = {
      creds: null,
      chats: null,
      messages: {},
      activeChat: null,
      typingCounter: {},
      typingData: {}
    };
    return _this;
  }

  var _proto = App.prototype;

  _proto.onConnect = function onConnect(creds) {
    this.setState({
      creds: creds
    });
    getChats(creds, function (data) {});
    this.props.onConnect && this.props.onConnect(creds);
  };

  _proto.onFailAuth = function onFailAuth(creds) {
    this.setState({
      creds: undefined
    });
    this.props.onFailAuth && this.props.onFailAuth(creds);
  };

  _proto.setActiveChat = function setActiveChat(chatId) {
    this.setState({
      activeChat: chatId
    });
    getMessages(this.state.creds, chatId, function (data) {});
  };

  _proto.onGetChats = function onGetChats(chats) {
    if (chats.length > 0) {
      this.setActiveChat(chats[0].id);
    }

    this.setState({
      chats: _.mapKeys(chats, 'id')
    });
    this.props.onGetChats && this.props.onGetChats(chats);
  };

  _proto.onNewChat = function onNewChat(chat) {
    var chats = this.state.chats;

    if (chats) {
      chats[chat.id] = chat;
      this.setState({
        chats: chats,
        activeChat: chat.id
      });
    }

    this.props.onNewChat && this.props.onNewChat(chat);
  };

  _proto.onEditChat = function onEditChat(chat) {
    var chats = this.state.chats;

    if (chats) {
      chats[chat.id] = chat;
      this.setState({
        chats: chats
      });
    }

    this.props.onEditChat && this.props.onEditChat(chat);
  };

  _proto.onDeleteChat = function onDeleteChat(chat) {
    var chats = this.state.chats;

    if (chats) {
      chats[chat.id] = undefined;
      this.setState({
        chats: chats
      });
    }

    this.props.onDeleteChat && this.props.onDeleteChat(chat);
  };

  _proto.onGetMessages = function onGetMessages(chatId, messages) {
    this.setState({
      messages: _.mapKeys(messages, 'id')
    });

    if (messages.length > 0) {
      var messageId = messages[messages.length - 1].id;
      readMessage(this.state.creds, this.state.activeChat, messageId, function (data) {});
    }

    this.props.onGetMessages && this.props.onGetMessages(chatId, messages);
  };

  _proto.onNewMessage = function onNewMessage(chatId, message) {
    if (chatId === this.state.activeChat) {
      var messages = this.state.messages;
      messages[message.id] = message;
      this.setState({
        messages: messages
      });
    }

    readMessage(this.state.creds, this.state.activeChat, message.id);
    this.props.onNewMessage && this.props.onNewMessage(chatId, message);
  };

  _proto.onEditMessage = function onEditMessage(chatId, message) {
    if (chatId === this.state.activeChat) {
      var messages = this.state.messages;
      messages[message.id] = message;
      this.setState({
        messages: messages
      });
    }

    this.props.onEditMessage && this.props.onEditMessage(chatId, message);
  };

  _proto.onDeleteMessage = function onDeleteMessage(chatId, message) {
    if (chatId === this.state.activeChat) {
      var messages = this.state.messages;
      messages[message.id] = undefined;
      this.setState({
        messages: messages
      });
    }

    this.props.onDeleteMessage && this.props.onDeleteMessage(chatId, message);
  };

  _proto.onTyping = function onTyping(chatId, person) {
    var _this2 = this;

    if (this.state.typingCounter[chatId] && this.state.typingCounter[chatId][person]) {
      var _extends2, _extends3;

      this.setState(_extends({}, this.state, {
        typingCounter: _extends({}, this.state.typingCounter, (_extends3 = {}, _extends3[chatId] = _extends({}, this.state.typingCounter[chatId], (_extends2 = {}, _extends2[person] = this.state.typingCounter[chatId][person] + 1, _extends2)), _extends3))
      }));
    } else {
      var _extends4, _extends5;

      this.setState(_extends({}, this.state, {
        typingCounter: _extends({}, this.state.typingCounter, (_extends5 = {}, _extends5[chatId] = _extends({}, this.state.typingCounter[chatId], (_extends4 = {}, _extends4[person] = 1, _extends4)), _extends5))
      }));
    }

    this.props.onTyping && this.props.onTyping(chatId, person);
    setTimeout(function () {
      var _extends6, _extends7;

      _this2.setState(_extends({}, _this2.state, {
        typingCounter: _extends({}, _this2.state.typingCounter, (_extends7 = {}, _extends7[chatId] = _extends({}, _this2.state.typingCounter[chatId], (_extends6 = {}, _extends6[person] = _this2.state.typingCounter[chatId][person] - 1, _extends6)), _extends7))
      }));
    }, 1250);
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    var _this3 = this;

    var _this$state = this.state,
        typingCounter = _this$state.typingCounter,
        typingData = _this$state.typingData;
    Object.keys(typingCounter).map(function (chat) {
      var typers = [];
      Object.keys(typingCounter[chat]).map(function (person) {
        if (typingCounter[chat][person] > 0) {
          typers.push(person);
        }
      });

      if (!typingData[chat] || typingData[chat].length !== typers.length) {
        var _extends8;

        _this3.setState(_extends({}, _this3.state, {
          typingData: _extends({}, _this3.state.typingData, (_extends8 = {}, _extends8[chat] = typers, _extends8))
        }));
      }
    });
  };

  _proto.render = function render() {
    var _this4 = this;

    var height = this.props.height;
    return /*#__PURE__*/React.createElement("div", {
      style: {
        textAlign: 'left',
        backgroundColor: 'white'
      }
    }, /*#__PURE__*/React.createElement(Socket, _extends({}, this.props, {
      onConnect: function onConnect(props) {
        return _this4.onConnect(props);
      },
      onFailAuth: function onFailAuth(props) {
        return _this4.onFailAuth(props);
      },
      onGetChats: function onGetChats(chats) {
        return _this4.onGetChats(chats);
      },
      onNewChat: function onNewChat(chat) {
        return _this4.onNewChat(chat);
      },
      onEditChat: function onEditChat(chat) {
        return _this4.onEditChat(chat);
      },
      onDeleteChat: function onDeleteChat(chat) {
        return _this4.onDeleteChat(chat);
      },
      onAddPerson: function onAddPerson(chat) {
        return _this4.onEditChat(chat);
      },
      onRemovePerson: function onRemovePerson(chat) {
        return _this4.onEditChat(chat);
      },
      onTyping: function onTyping(chatId, person) {
        return _this4.onTyping(chatId, person);
      },
      onGetMessages: function onGetMessages(chatId, messages) {
        return _this4.onGetMessages(chatId, messages);
      },
      onNewMessage: function onNewMessage(chatId, message) {
        return _this4.onNewMessage(chatId, message);
      },
      onEditMessage: function onEditMessage(chatId, message) {
        return _this4.onEditMessage(chatId, message);
      },
      onDeleteMessage: function onDeleteMessage(chatId, message) {
        return _this4.onDeleteMessage(chatId, message);
      }
    })), /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Col, {
      xs: 0,
      sm: 3,
      style: {
        height: height ? height : '',
        overflow: 'scroll'
      }
    }, /*#__PURE__*/React.createElement(ChatList, {
      creds: this.state.creds,
      chats: this.state.chats,
      activeChat: this.state.activeChat,
      onChatClick: function onChatClick(chatId) {
        return _this4.setActiveChat(chatId);
      },
      renderChatCard: this.props.renderChatCard,
      renderNewChatForm: this.props.renderNewChatForm
    })), /*#__PURE__*/React.createElement(Col, {
      xs: 12,
      sm: 6,
      style: {
        height: height ? height : '',
        overflow: 'scroll'
      }
    }, /*#__PURE__*/React.createElement(ChatList$1, {
      creds: this.state.creds,
      chats: this.state.chats,
      chatId: this.state.activeChat,
      messages: this.state.messages,
      typingData: this.state.typingData,
      renderChatHeader: this.props.renderChatHeader,
      renderMessageBubble: this.props.renderMessageBubble,
      renderIsTyping: this.props.renderIsTyping,
      renderNewMessageForm: this.props.renderNewMessageForm
    })), /*#__PURE__*/React.createElement(Col, {
      xs: 0,
      sm: 3,
      style: {
        height: height ? height : '',
        overflow: 'scroll'
      }
    }, /*#__PURE__*/React.createElement(ChatSettingsContainer, {
      creds: this.state.creds,
      chat: this.state.chats && this.state.chats[this.state.activeChat],
      renderChatSettings: this.props.renderChatSettings,
      renderPeopleSettings: this.props.renderPeopleSettings,
      renderPhotosSettings: this.props.renderPhotosSettings,
      renderOptionsSettings: this.props.renderOptionsSettings
    }))));
  };

  return App;
}(Component);

var SocketConnector = function SocketConnector(props) {
  if (props.hideUI) {
    return /*#__PURE__*/React.createElement(Socket, props);
  } else {
    return /*#__PURE__*/React.createElement(App, props);
  }
};

function isTyping(props, chatId, callback) {
  axios.post(getApiUrl(props) + "/chats/" + chatId + "/typing/", {}, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {});
}

function addPerson(props, chatId, userName, callback) {
  axios.post(getApiUrl(props) + "/chats/" + chatId + "/people/", {
    username: userName
  }, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('New Person Error', error);
  });
}

function removePerson(props, chatId, userName, callback) {
  axios.put(getApiUrl(props) + "/chats/" + chatId + "/people/", {
    username: userName
  }, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Delete Person Error', error);
  });
}

function getOtherPeople(props, chatId, successCallback, errorCallback) {
  axios.get(getApiUrl(props) + "/chats/" + chatId + "/others/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    props.onGetOtherPeople && props.onGetOtherPeople(chatId, response.data);
    successCallback && successCallback(chatId, response.data);
  })["catch"](function (error) {
    console.log('Fetch Other People Error', error);
    errorCallback && errorCallback();
  });
}

function getMyData(props, callback) {
  axios.get(getApiUrl(props) + "/chats/me/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Get Myself Error', error);
  });
}

function editMyData(props, data, callback) {
  axios.patch(getApiUrl(props) + "/chats/me/", data, {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Edit Myself Error', error);
  });
}

function leaveChat(props, chatId, callback) {
  axios["delete"](getApiUrl(props) + "/chats/" + chatId + "/people/", {
    headers: {
      "Public-Key": props.publicKey ? props.publicKey : props.projectID,
      "User-Name": props.userName,
      "User-Secret": props.userPassword ? props.userPassword : props.userSecret
    }
  }).then(function (response) {
    callback && callback(response.data);
  })["catch"](function (error) {
    console.log('Delete Person Error', error);
  });
}

var ChatEngine = function ChatEngine(props) {
  return /*#__PURE__*/React.createElement(SocketConnector, props);
};

export { ChatEngine, addPerson, deleteChat, deleteMessage, editChat, editMessage, editMyData, getChats, getMessages, getMyData, getOtherPeople, isTyping, leaveChat, newChat, removePerson, sendMessage };
